iree/compiler/Dialect/HAL/IR/HALDialect.cpp

```diff
@@ -54,10 +56,15 @@ class HALToVMConversionInterface : public VMConversionDialectInterface {
  public:
   using VMConversionDialectInterface::VMConversionDialectInterface;
 
-  OwningModuleRef getVMImportModule() const override {
-    return mlir::parseSourceString(
-        StringRef(hal_imports_create()->data, hal_imports_create()->size),
-        getDialect()->getContext());
+  mlir::ModuleOp getVMImportModule() const override {
+    static OwningModuleRef importModuleRef;
+    static std::once_flag parseOnce;
```

Comment:
Seems worth a comment explaining that this is correct (it doesn't make sense to have multiple import modules per component). Could comment on why this also helps performance if you want.

(same in other files that repeat this pattern)

Reply:
made the code clearer and hid it in the base class